storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +274,228 @@ func (s *commitLogSource) Read(
+	return bootstrapResult
+}
+
+func (s *commitLogSource) mergeShard(
+	unmergedShard encodersAndRanges,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (shardResult result.ShardResult, numShardEmptyErrs int, numErrs int) {
+
+	for _, unmergedBlocks := range unmergedShard.encodersBySeries {
+		seriesBlocks, numSeriesEmptyErrs, numSeriesErrs := s.mergeSeries(
+			unmergedBlocks,
+			blocksPool,
+			multiReaderIteratorPool,
+			encoderPool,
+			blopts,
+		)
+
+		if seriesBlocks != nil && seriesBlocks.Len() > 0 {
+			if shardResult == nil {
+				shardResult = result.NewShardResult(len(unmergedShard.encodersBySeries), s.opts.ResultOptions())
+			}
+			shardResult.AddSeries(unmergedBlocks.id, seriesBlocks)
+		}
+
+		numShardEmptyErrs += numSeriesEmptyErrs
+		numErrs += numSeriesErrs
+	}
+	return shardResult, numShardEmptyErrs, numErrs
+}
+
+func (s *commitLogSource) mergeSeries(
+	unmergedBlocks encodersByTime,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (seriesBlocks block.DatabaseSeriesBlocks, numEmptyErrs int, numErrs int) {
+
+	for startNano, encoders := range unmergedBlocks.encoders {
+		start := startNano.ToTime()
+
+		if len(encoders) == 0 {
+			numEmptyErrs++
+			continue
+		}
+
+		if len(encoders) == 1 {
+			pooledBlock := blocksPool.Get()
+			pooledBlock.Reset(start, encoders[0].enc.Discard())
+			if seriesBlocks == nil {
+				seriesBlocks = block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
 			}
-			if blocks.Len() > 0 {
-				shardResult.AddSeries(unmergedBlocks.id, blocks)
+			seriesBlocks.AddBlock(pooledBlock)
+			continue
+		}
+
+		// Convert encoders to readers so we can use iteration helpers
+		readers := encoders.newReaders()
+		iter := multiReaderIteratorPool.Get()
+		iter.Reset(readers)
+
+		var err error
+		enc := encoderPool.Get()
+		enc.Reset(start, blopts.DatabaseBlockAllocSize())
+		for iter.Next() {
+			dp, unit, annotation := iter.Current()
+			encodeErr := enc.Encode(dp, unit, annotation)
+			if encodeErr != nil {
+				err = encodeErr
+				numErrs++
+				break
 			}
 		}
-		if len(shardResult.AllSeries()) > 0 {
-			bootstrapResult.Add(shard, shardResult, nil)
+
+		if iterErr := iter.Err(); iterErr != nil {
+			if err == nil {
+				err = iter.Err()
+			}
+			numErrs++
+		}
+
+		// Automatically returns iter to the pool
+		iter.Close()
+		encoders.close()
+		readers.close()
+
+		if err != nil {
+			continue
 		}
+
+		pooledBlock := blocksPool.Get()
+		pooledBlock.Reset(start, enc.Discard())
+		if seriesBlocks == nil {
+			seriesBlocks = block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
+		}
+		seriesBlocks.AddBlock(pooledBlock)
 	}
-	if errs > 0 {
-		s.log.Errorf("error bootstrapping from commit log: %d merge out of order errors", errs)
+	return seriesBlocks, numEmptyErrs, numErrs
```

Comment:
Weird to be using named return vars and overriding them instead of a naked return. Could you pick one convention

Reply:
I actually kind of like this pattern. Naked returns are generally frowned upon for long functions, but I do like being able to name the returns because it makes the function signature cleaner

Reply:
Can change it to a naked return if you feel strongly

Reply:
Yeah call me c99 bent but I don't like named returns ever (and naked returns as a result). Could you declare the vars instead of using named returns, and then leave the return as is? 

The concern is if someone adds a return to early terminate the function without realising the lifecycle of the named return vars